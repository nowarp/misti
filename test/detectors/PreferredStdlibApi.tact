message Msg {
    whatever: Slice as remaining;
}

contract Test {
    receive(msg: Msg) {
        let pkg: Slice = msg.whatever;
        let _seqno: Int = pkg.loadInt(32);
        let mode: Int = pkg.loadInt(8);
        let body: Cell = pkg.loadRef();
        nativeSendMessage(body, mode); // Bad: prefer `send`
    }

    fun getInt(): Int {
        return nativeRandom(); // Bad: prefer `randomInt`
    }
}
