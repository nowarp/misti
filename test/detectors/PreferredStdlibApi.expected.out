[INFO] PreferredStdlibApi: nativeSendMessage has a safer alternative: send
test/detectors/PreferredStdlibApi.tact:11:9:
  10 |         let body: Cell = pkg.loadRef();
> 11 |         nativeSendMessage(body, mode); // Bad: prefer `send`
               ^
  12 |     }
Prefer `send` to make the call more explicit and reduce low-level operations
Help: nativeSendMessage should be replaced with a safer alternative: send
See: https://nowarp.io/tools/misti/docs/detectors/PreferredStdlibApi

[INFO] PreferredStdlibApi: nativeRandom has a safer alternative: randomInt
test/detectors/PreferredStdlibApi.tact:15:16:
  14 |     fun getInt(): Int {
> 15 |         return nativeRandom(); // Bad: prefer `randomInt`
                      ^
  16 |     }
Prefer `randomInt` since `nativeRandom` requires additional initialization of PRG before use
Help: nativeRandom should be replaced with a safer alternative: randomInt
See: https://nowarp.io/tools/misti/docs/detectors/PreferredStdlibApi